File: src/main/java/hzt/aoc/day07/Part2HandyHaversacks.java
package hzt.aoc.day07;

import hzt.collections.MapX;

public class Part2HandyHaversacks extends Day07Challenge {

    public Part2HandyHaversacks() {
        super("part 2",
                "How many individual bags are required inside your single shiny gold bag?");
    }

    @Override
    protected long solveByRules(MapX<String, Bag> bags) {
        // We counted the target bag, reduce count by 1.
        return bags.isNotEmpty() ? (countInnerBagsRecursive(bags, bags.get(SHINY_GOLD)) - 1) : 0;
    }

    private static long countInnerBagsRecursive(MapX<String, Bag> bags, Bag bag) {
        return bag.getChildBagColorsToAmounts().entrySet()
                .sumOfLongs(entry -> entry.getValue() * countInnerBagsRecursive(bags, bags.get(entry.getKey()))) + 1;
    }

    @Override
    public String getMessage(String numberOfBags) {
        return String.format("The number of individual bags required inside the %s bag: %s%n", SHINY_GOLD, numberOfBags);
    }
}

File: src/main/kotlin/hzt/aoc/day07/Part2HandyHaversacks.kt
package hzt.aoc.day07

class Part2HandyHaversacks : Day07Challenge(
    "part 2",
    "How many individual bags are required inside your single shiny gold bag"
) {
    override fun solveByRules(bags: Map<String, Bag>): Long =
        if (bags.isNotEmpty()) countInnerBagsRecursive(bags, bags[SHINY_GOLD]) - 1 else 0 // We counted the target bag, reduce count by 1.

    private fun countInnerBagsRecursive(bags: Map<String, Bag>, bag: Bag?): Long {
        var accumulator: Long = 1
        accumulator += bag?.childBagColorsToAmounts?.entries?.asSequence()
            ?.map { (name, count) -> count * countInnerBagsRecursive(bags, bags[name]) }
            ?.sum() ?: 0
        return accumulator
    }

    public override fun getMessage(result: String): String = String.format(
        "The number of individual bags required inside the %s bag: %s%n", SHINY_GOLD, result
    )
}
